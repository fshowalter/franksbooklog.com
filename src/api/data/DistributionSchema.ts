import { z } from "zod";

/**
 * Zod schema for distribution data used in statistics.
 * Represents a named category with an associated count value.
 */
export const DistributionSchema = z.object({
  count: z.number(),
  name: z.string(),
});
