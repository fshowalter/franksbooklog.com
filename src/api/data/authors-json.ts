import { promises as fs } from "node:fs";
import { z } from "zod";

import { perfLogger } from "~/utils/performanceLogger";

import { getContentPath } from "./utils/getContentPath";
import { nullableString } from "./utils/nullable";
import { WorkKindSchema } from "./WorkKindSchema";

const authorsJsonDirectory = getContentPath("data", "authors");

/**
 * Zod schema for author information within a work's metadata.
 * Represents an individual author with their basic information.
 */
const WorkAuthorSchema = z
  .object({
    name: z.string(),
    notes: nullableString(),
    slug: z.string(),
    sortName: z.string(),
  })
  .transform(({ name, notes, slug, sortName }) => {
    // fix zod making anything with undefined optional
    return { name, notes, slug, sortName };
  });

/**
 * Zod schema for work information within an author's reviewed works.
 * Contains comprehensive metadata about a book/work including review details.
 */
const WorkSchema = z.object({
  authors: z.array(WorkAuthorSchema),
  grade: z.string(),
  gradeValue: z.number(),
  includedInSlugs: z.array(z.string()),
  kind: WorkKindSchema,
  reviewDate: z.string(),
  reviewSequence: z.string(),
  reviewYear: z.string(),
  slug: z.string(),
  sortTitle: z.string(),
  subtitle: nullableString(),
  title: z.string(),
  workYear: z.string(),
});

/**
 * Zod schema for complete author data from JSON files.
 * Includes author metadata and all their reviewed works.
 */
const AuthorJsonSchema = z.object({
  name: z.string(),
  reviewedWorks: z.array(WorkSchema),
  slug: z.string(),
  sortName: z.string(),
});

/**
 * Author data structure from JSON files in the content directory
 */
export type AuthorJson = z.infer<typeof AuthorJsonSchema>;

// Cache at data layer - lazy caching for better build performance
let cachedAuthorsJson: AuthorJson[];

// Enable caching during builds but not in dev mode
const ENABLE_CACHE = !import.meta.env.DEV;

/**
 * Loads and validates all author JSON files from the content directory.
 * Authors are parsed with Zod schemas and cached during builds.
 *
 * @returns Promise resolving to array of validated author data
 */
export async function allAuthorsJson(): Promise<AuthorJson[]> {
  return await perfLogger.measure("allAuthorsJson", async () => {
    if (ENABLE_CACHE && cachedAuthorsJson) {
      return cachedAuthorsJson;
    }

    const authors = await parseAllAuthorsJson();

    if (ENABLE_CACHE) {
      cachedAuthorsJson = authors;
    }

    return authors;
  });
}

/**
 * Internal function to parse all author JSON files from the file system.
 * Reads the authors directory and validates each JSON file against the schema.
 *
 * @returns Promise resolving to array of parsed and validated author data
 * @throws ZodError if any JSON file doesn't match the expected schema
 */
async function parseAllAuthorsJson() {
  return await perfLogger.measure("parseAllAuthorsJson", async () => {
    const dirents = await fs.readdir(authorsJsonDirectory, {
      withFileTypes: true,
    });

    return Promise.all(
      dirents
        .filter((entry) => !entry.isDirectory() && entry.name.endsWith(".json"))
        .map(async (entry) => {
          const fileContents = await fs.readFile(
            `${authorsJsonDirectory}/${entry.name}`,
            "utf8",
          );

          const json = JSON.parse(fileContents) as unknown;
          return AuthorJsonSchema.parse(json);
        }),
    );
  });
}
