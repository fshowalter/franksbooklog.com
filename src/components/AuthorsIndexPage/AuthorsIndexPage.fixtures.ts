export default {
  author: {
    nodes: [
      {
        name: "Dave Draper",
        sortName: "Draper, Dave",
        slug: "dave-draper",
        workCount: 1,
        reviewedWorkCount: 1,
        avatar: null,
      },
      {
        name: "Stephen King",
        sortName: "King, Stephen",
        slug: "stephen-king",
        workCount: 1,
        reviewedWorkCount: 1,
        avatar: null,
      },
      {
        name: "Richard Laymon",
        sortName: "Laymon, Richard",
        slug: "richard-laymon",
        workCount: 1,
        reviewedWorkCount: 1,
        avatar: null,
      },
      {
        name: "Brad Schoenfeld",
        sortName: "Schoenfeld, Brad",
        slug: "brad-schoenfeld",
        workCount: 1,
        reviewedWorkCount: 1,
        avatar: null,
      },
    ],
  },
};
