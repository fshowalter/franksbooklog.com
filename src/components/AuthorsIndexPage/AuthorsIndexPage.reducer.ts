import applyFilters from "../../utils/apply-filters";
import { sortNumberDesc, sortStringAsc } from "../../utils/sort-utils";
import type { Author } from "./AuthorsIndexPage";

export enum ActionType {
  FILTER_NAME = "FILTER_NAME",
  SORT = "SORT",
}

export type SortValue = "name" | "reviews" | "works";

function sortAuthors(authors: Author[], sortOrder: SortValue) {
  const sortMap: Record<SortValue, (a: Author, b: Author) => number> = {
    name: (a, b) => sortStringAsc(a.sortName, b.sortName),
    works: (a, b) => sortNumberDesc(a.workCount, b.workCount),
    reviews: (a, b) => sortNumberDesc(a.reviewedWorkCount, b.reviewedWorkCount),
  };

  const comparer = sortMap[sortOrder];

  return authors.sort(comparer);
}

/**
 * The page state.
 */
type State = {
  /** All possible reviews. */
  allAuthors: Author[];
  /** People matching the current filters. */
  filteredAuthors: Author[];
  /** The active filters. */
  filters: Record<string, (author: Author) => boolean>;
  /** The active sort value. */
  sortValue: SortValue;
};

export function initState({ authors }: { authors: Author[] }): State {
  return {
    allAuthors: authors,
    filteredAuthors: authors,
    filters: {},
    sortValue: "name",
  };
}

/** Action to filter by title. */
interface FilterNameAction {
  type: ActionType.FILTER_NAME;
  /** The value to filter on. */
  value: string;
}

interface SortAction {
  type: ActionType.SORT;
  /** The sorter to apply. */
  value: SortValue;
}

export type Action = FilterNameAction | SortAction;

/**
 * Applies the given action to the given state, returning a new State object.
 * @param state The current state.
 * @param action The action to apply.
 */
export function reducer(state: State, action: Action): State {
  let filters;
  let filteredAuthors;

  switch (action.type) {
    case ActionType.FILTER_NAME: {
      const regex = new RegExp(action.value, "i");
      filters = {
        ...state.filters,
        name: (author: Author) => {
          return regex.test(author.name);
        },
      };
      filteredAuthors = sortAuthors(
        applyFilters<Author>({
          collection: state.allAuthors,
          filters,
        }),
        state.sortValue
      );
      return {
        ...state,
        filters,
        filteredAuthors,
      };
    }
    case ActionType.SORT: {
      filteredAuthors = sortAuthors(state.filteredAuthors, action.value);
      return {
        ...state,
        sortValue: action.value,
        filteredAuthors,
      };
    }
    // no default
  }
}
