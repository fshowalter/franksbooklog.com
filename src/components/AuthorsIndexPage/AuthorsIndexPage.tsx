import { graphql, Link } from "gatsby";
import { GatsbyImage, IGatsbyImageData } from "gatsby-plugin-image";
import { useReducer } from "react";
import DebouncedInput from "../DebouncedInput";
import Fieldset from "../Fieldset";
import FilterPageHeader from "../FilterPageHeader";
import HeadBuilder from "../HeadBuilder";
import Layout from "../Layout";
import { SelectField } from "../SelectField";
import {
  containerCss,
  defaultImageCss,
  filtersCss,
  leftCss,
  listCss,
  listItemAvatarCss,
  listItemCss,
  listItemLinkCss,
  listItemTitleCss,
  pageHeaderCss,
  percentBackgroundCss,
  percentProgressCss,
  progressRingCss,
  progressStatsCss,
  rightCss,
} from "./AuthorsIndexPage.module.scss";
import {
  ActionType,
  initState,
  reducer,
  SortValue,
} from "./AuthorsIndexPage.reducer";

function Progress({ author }: { author: Author }): JSX.Element {
  const percent = author.reviewedWorkCount / author.workCount;
  const circumference = 17.5 * 2 * Math.PI;

  return (
    <svg
      viewBox="0 0 36 36"
      className={progressRingCss}
      preserveAspectRatio="none"
    >
      <g id="circles" strokeWidth="1">
        <circle
          r="17.5"
          cx="18"
          cy="18"
          fill="none"
          className={percentBackgroundCss}
        />
        <circle
          r="17.5"
          cx="18"
          cy="18"
          fill="none"
          strokeDasharray={`${circumference} ${circumference}`}
          strokeDashoffset={circumference - percent * circumference}
          className={percentProgressCss}
        />
      </g>
    </svg>
  );
}

function ListItem({ author }: { author: Author }): JSX.Element {
  if (author.avatar) {
    return (
      <li className={listItemCss}>
        <Link className={listItemLinkCss} to={`/shelf/authors/${author.slug}/`}>
          {author.avatar && (
            <GatsbyImage
              image={author.avatar.childImageSharp.gatsbyImageData}
              className={listItemAvatarCss}
              alt={`An image of ${author.name}`}
            />
          )}
          <Progress author={author} />
        </Link>
        <div className={listItemTitleCss}>
          <Link to={`/shelf/authors/${author.slug}/`}>{author.name}</Link>
        </div>
        <div className={progressStatsCss}>
          {author.reviewedWorkCount} / {author.workCount}
        </div>
      </li>
    );
  }
  return (
    <li className={listItemCss}>
      <div className={listItemLinkCss}>
        <Link className={listItemLinkCss} to={`/shelf/authors/${author.slug}/`}>
          <svg
            className={`${listItemAvatarCss} ${defaultImageCss}`}
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 16 16"
            fill="currentColor"
          >
            <path
              clipRule="evenodd"
              d="M16 8A8 8 0 110 8a8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zM8 9a5 5 0 00-4.546 2.916A5.986 5.986 0 008 14a5.986 5.986 0 004.546-2.084A5 5 0 008 9z"
              fillRule="evenodd"
            />
          </svg>
          <Progress author={author} />
        </Link>
      </div>
      <div className={listItemTitleCss}>
        <Link to={`/shelf/authors/${author.slug}/`}>{author.name}</Link>
      </div>
      <div className={progressStatsCss}>
        {author.reviewedWorkCount} / {author.workCount}
      </div>
    </li>
  );
}

export function Head(): JSX.Element {
  return (
    <HeadBuilder
      pageTitle={`Authors`}
      description={`A sortable and filterable list of shelf authors.`}
      image={null}
      article={false}
    />
  );
}

/**
 * Renders an index page for authors.
 */
export default function AuthorIndexPage({
  data,
}: {
  data: PageQueryResult;
}): JSX.Element {
  const [state, dispatch] = useReducer(
    reducer,
    {
      authors: [...data.author.nodes],
    },
    initState
  );

  return (
    <Layout>
      <main className={containerCss}>
        <div className={leftCss}>
          <FilterPageHeader
            className={pageHeaderCss}
            heading={`Authors`}
            tagline={
              <q>
                There is nothing to writing. All you do is sit down at a
                typewriter and bleed.
              </q>
            }
            breadcrumb={<Link to="/shelf/">Shelf</Link>}
          />
          <div className={filtersCss}>
            <Fieldset legend="Filter & Sort">
              <DebouncedInput
                label="Name"
                placeholder="Enter all or part of a name"
                onChange={(value) =>
                  dispatch({ type: ActionType.FILTER_NAME, value })
                }
              />
              <SelectField
                value={state.sortValue}
                label="Order By"
                onChange={(e) =>
                  dispatch({
                    type: ActionType.SORT,
                    value: e.target.value as SortValue,
                  })
                }
              >
                <option value="name">Name</option>
                <option value="works">Work Count</option>
                <option value="reviews">Review Count</option>
              </SelectField>
            </Fieldset>
          </div>
        </div>
        <div className={rightCss}>
          <ul data-testid="author-list" className={listCss}>
            {state.filteredAuthors.map((author) => {
              return <ListItem key={author.slug} author={author} />;
            })}
          </ul>
        </div>
      </main>
    </Layout>
  );
}

export interface Author {
  name: string;
  slug: string;
  workCount: number;
  reviewedWorkCount: number;
  sortName: string;
  avatar: null | {
    childImageSharp: {
      gatsbyImageData: IGatsbyImageData;
    };
  };
}

interface PageQueryResult {
  author: {
    nodes: Author[];
  };
}

export const pageQuery = graphql`
  {
    author: allAuthorsJson(
      filter: { reviewed: { eq: true } }
      sort: { sort_name: ASC }
    ) {
      nodes {
        name
        sortName: sort_name
        slug
        workCount
        reviewedWorkCount
        avatar {
          childImageSharp {
            gatsbyImageData(
              layout: CONSTRAINED
              formats: [JPG, AVIF]
              quality: 80
              width: 160
              height: 160
              placeholder: TRACED_SVG
            )
          }
        }
      }
    }
  }
`;
