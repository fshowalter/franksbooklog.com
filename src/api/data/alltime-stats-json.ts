import { promises as fs } from "node:fs";
import { z } from "zod";

import { perfLogger } from "~/utils/performanceLogger";

import { DistributionSchema } from "./DistributionSchema";
import { MostReadAuthorSchema } from "./MostReadAuthorSchema";
import { getContentPath } from "./utils/getContentPath";

const alltimeStatsFile = getContentPath("data", "all-time-stats.json");

/**
 * Extended distribution schema for grade data that includes a sort value.
 * Used specifically for organizing review grades in proper order.
 */
const GradeDistributionSchema = DistributionSchema.extend({
  sortValue: z.number(),
});

/**
 * Zod schema for all-time statistics data loaded from JSON.
 * Validates the structure of comprehensive reading statistics across all years.
 */
const AlltimeStatsJsonSchema = z.object({
  bookCount: z.number(),
  decadeDistribution: z.array(DistributionSchema),
  editionDistribution: z.array(DistributionSchema),
  gradeDistribution: z.array(GradeDistributionSchema),
  kindDistribution: z.array(DistributionSchema),
  mostReadAuthors: z.array(MostReadAuthorSchema),
  reviewCount: z.number(),
  workCount: z.number(),
});

/**
 * Type representing all-time reading statistics data.
 * Contains comprehensive metrics including counts, distributions, and top authors.
 */
export type AlltimeStatsJson = z.infer<typeof AlltimeStatsJsonSchema>;

/**
 * Loads and validates all-time reading statistics from JSON file.
 * Provides comprehensive lifetime reading metrics including book counts,
 * grade distributions, most read authors, and categorical breakdowns.
 *
 * @returns Promise resolving to validated all-time statistics data
 * @throws ZodError if the JSON data doesn't match the expected schema
 *
 * @example
 * ```typescript
 * const stats = await alltimeStatsJson();
 * console.log(`Total books read: ${stats.bookCount}`);
 * console.log(`Top author: ${stats.mostReadAuthors[0].name}`);
 * ```
 */
export async function alltimeStatsJson(): Promise<AlltimeStatsJson> {
  return await perfLogger.measure("alltimeStatsJson", async () => {
    const json = await fs.readFile(alltimeStatsFile, "utf8");
    const data = JSON.parse(json) as unknown[];

    return AlltimeStatsJsonSchema.parse(data);
  });
}
