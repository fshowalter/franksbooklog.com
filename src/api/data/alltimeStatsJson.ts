import { promises as fs } from "node:fs";
import { z } from "zod";

import { DistributionSchema } from "./DistributionSchema";
import { MostReadAuthorSchema } from "./MostReadAuthorSchema";
import { getContentPath } from "./utils/getContentPath";

const alltimeStatsFile = getContentPath("data", "all-time-stats.json");

const GradeDistributionSchema = DistributionSchema.extend({
  sortValue: z.number(),
});

const AlltimeStatsJsonSchema = z.object({
  bookCount: z.number(),
  decadeDistribution: z.array(DistributionSchema),
  editionDistribution: z.array(DistributionSchema),
  gradeDistribution: z.array(GradeDistributionSchema),
  kindDistribution: z.array(DistributionSchema),
  mostReadAuthors: z.array(MostReadAuthorSchema),
  reviewCount: z.number(),
  workCount: z.number(),
});

export type AlltimeStatsJson = z.infer<typeof AlltimeStatsJsonSchema>;

export async function alltimeStatsJson(): Promise<AlltimeStatsJson> {
  const json = await fs.readFile(alltimeStatsFile, "utf8");
  const data = JSON.parse(json) as unknown[];

  return AlltimeStatsJsonSchema.parse(data);
}
