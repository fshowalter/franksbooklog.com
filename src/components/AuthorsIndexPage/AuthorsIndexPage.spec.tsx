// eslint-disable-next-line import/no-extraneous-dependencies
import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import AuthorsIndexPage from "./AuthorsIndexPage";
import data from "./AuthorsIndexPage.fixtures";

describe("/shelf/authors/", () => {
  it("can filter by name", async () => {
    expect.hasAssertions();
    render(<AuthorsIndexPage data={data} />);

    act(() => {
      jest.useFakeTimers(); // For the debouced input
      userEvent.type(screen.getByLabelText("Name"), "Stephen King");
      jest.runOnlyPendingTimers(); // Flush the delay
      jest.useRealTimers();
    });

    await waitFor(() => {
      expect(screen.getByTestId("author-list")).toMatchSnapshot();
    });
  });

  it("can sort by name", () => {
    render(<AuthorsIndexPage data={data} />);

    userEvent.selectOptions(screen.getByLabelText("Order By"), "Name");

    expect(screen.getByTestId("author-list")).toMatchSnapshot();
  });

  it("can sort by review count", () => {
    render(<AuthorsIndexPage data={data} />);

    userEvent.selectOptions(screen.getByLabelText("Order By"), "Review Count");

    expect(screen.getByTestId("author-list")).toMatchSnapshot();
  });

  // Helmet uses requestAnimationFrame to ensure DOM is synced.
  // https://github.com/nfl/react-helmet/blob/master/test/HelmetDeclarativeTest.js
  // eslint-disable-next-line jest/no-done-callback
  it("sets page title", (done) => {
    expect.hasAssertions();
    render(<AuthorsIndexPage data={data} />);

    requestAnimationFrame(() => {
      expect(document.title).toStrictEqual("Shelf Authors");
      done();
    });
  });
});
