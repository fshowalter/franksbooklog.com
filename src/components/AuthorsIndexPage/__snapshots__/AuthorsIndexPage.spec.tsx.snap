// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/shelf/authors/ can filter by name 1`] = `
<ul
  class="listCss"
  data-testid="author-list"
>
  <li
    class="listItemCss"
  >
    <div
      class="listItemLinkCss"
    >
      <a
        class="listItemLinkCss"
        href="/shelf/authors/stephen-king/"
      >
        <svg
          class="listItemAvatarCss defaultImageCss"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16 8A8 8 0 110 8a8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zM8 9a5 5 0 00-4.546 2.916A5.986 5.986 0 008 14a5.986 5.986 0 004.546-2.084A5 5 0 008 9z"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          class="progressRingCss"
          preserveAspectRatio="none"
          viewBox="0 0 36 36"
        >
          <g
            id="circles"
            stroke-width="1"
          >
            <circle
              class="percentBackgroundCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
            />
            <circle
              class="percentProgressCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
              stroke-dasharray="109.95574287564276 109.95574287564276"
              stroke-dashoffset="0"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="listItemTitleCss"
    >
      <a
        href="/shelf/authors/stephen-king/"
      >
        Stephen King
      </a>
    </div>
    <div
      class="progressStatsCss"
    >
      1
       / 
      1
    </div>
  </li>
</ul>
`;

exports[`/shelf/authors/ can sort by name 1`] = `
<ul
  class="listCss"
  data-testid="author-list"
>
  <li
    class="listItemCss"
  >
    <div
      class="listItemLinkCss"
    >
      <a
        class="listItemLinkCss"
        href="/shelf/authors/dave-draper/"
      >
        <svg
          class="listItemAvatarCss defaultImageCss"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16 8A8 8 0 110 8a8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zM8 9a5 5 0 00-4.546 2.916A5.986 5.986 0 008 14a5.986 5.986 0 004.546-2.084A5 5 0 008 9z"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          class="progressRingCss"
          preserveAspectRatio="none"
          viewBox="0 0 36 36"
        >
          <g
            id="circles"
            stroke-width="1"
          >
            <circle
              class="percentBackgroundCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
            />
            <circle
              class="percentProgressCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
              stroke-dasharray="109.95574287564276 109.95574287564276"
              stroke-dashoffset="0"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="listItemTitleCss"
    >
      <a
        href="/shelf/authors/dave-draper/"
      >
        Dave Draper
      </a>
    </div>
    <div
      class="progressStatsCss"
    >
      1
       / 
      1
    </div>
  </li>
  <li
    class="listItemCss"
  >
    <div
      class="listItemLinkCss"
    >
      <a
        class="listItemLinkCss"
        href="/shelf/authors/stephen-king/"
      >
        <svg
          class="listItemAvatarCss defaultImageCss"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16 8A8 8 0 110 8a8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zM8 9a5 5 0 00-4.546 2.916A5.986 5.986 0 008 14a5.986 5.986 0 004.546-2.084A5 5 0 008 9z"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          class="progressRingCss"
          preserveAspectRatio="none"
          viewBox="0 0 36 36"
        >
          <g
            id="circles"
            stroke-width="1"
          >
            <circle
              class="percentBackgroundCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
            />
            <circle
              class="percentProgressCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
              stroke-dasharray="109.95574287564276 109.95574287564276"
              stroke-dashoffset="0"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="listItemTitleCss"
    >
      <a
        href="/shelf/authors/stephen-king/"
      >
        Stephen King
      </a>
    </div>
    <div
      class="progressStatsCss"
    >
      1
       / 
      1
    </div>
  </li>
  <li
    class="listItemCss"
  >
    <div
      class="listItemLinkCss"
    >
      <a
        class="listItemLinkCss"
        href="/shelf/authors/richard-laymon/"
      >
        <svg
          class="listItemAvatarCss defaultImageCss"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16 8A8 8 0 110 8a8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zM8 9a5 5 0 00-4.546 2.916A5.986 5.986 0 008 14a5.986 5.986 0 004.546-2.084A5 5 0 008 9z"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          class="progressRingCss"
          preserveAspectRatio="none"
          viewBox="0 0 36 36"
        >
          <g
            id="circles"
            stroke-width="1"
          >
            <circle
              class="percentBackgroundCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
            />
            <circle
              class="percentProgressCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
              stroke-dasharray="109.95574287564276 109.95574287564276"
              stroke-dashoffset="0"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="listItemTitleCss"
    >
      <a
        href="/shelf/authors/richard-laymon/"
      >
        Richard Laymon
      </a>
    </div>
    <div
      class="progressStatsCss"
    >
      1
       / 
      1
    </div>
  </li>
  <li
    class="listItemCss"
  >
    <div
      class="listItemLinkCss"
    >
      <a
        class="listItemLinkCss"
        href="/shelf/authors/brad-schoenfeld/"
      >
        <svg
          class="listItemAvatarCss defaultImageCss"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16 8A8 8 0 110 8a8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zM8 9a5 5 0 00-4.546 2.916A5.986 5.986 0 008 14a5.986 5.986 0 004.546-2.084A5 5 0 008 9z"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          class="progressRingCss"
          preserveAspectRatio="none"
          viewBox="0 0 36 36"
        >
          <g
            id="circles"
            stroke-width="1"
          >
            <circle
              class="percentBackgroundCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
            />
            <circle
              class="percentProgressCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
              stroke-dasharray="109.95574287564276 109.95574287564276"
              stroke-dashoffset="0"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="listItemTitleCss"
    >
      <a
        href="/shelf/authors/brad-schoenfeld/"
      >
        Brad Schoenfeld
      </a>
    </div>
    <div
      class="progressStatsCss"
    >
      1
       / 
      1
    </div>
  </li>
</ul>
`;

exports[`/shelf/authors/ can sort by review count 1`] = `
<ul
  class="listCss"
  data-testid="author-list"
>
  <li
    class="listItemCss"
  >
    <div
      class="listItemLinkCss"
    >
      <a
        class="listItemLinkCss"
        href="/shelf/authors/dave-draper/"
      >
        <svg
          class="listItemAvatarCss defaultImageCss"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16 8A8 8 0 110 8a8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zM8 9a5 5 0 00-4.546 2.916A5.986 5.986 0 008 14a5.986 5.986 0 004.546-2.084A5 5 0 008 9z"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          class="progressRingCss"
          preserveAspectRatio="none"
          viewBox="0 0 36 36"
        >
          <g
            id="circles"
            stroke-width="1"
          >
            <circle
              class="percentBackgroundCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
            />
            <circle
              class="percentProgressCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
              stroke-dasharray="109.95574287564276 109.95574287564276"
              stroke-dashoffset="0"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="listItemTitleCss"
    >
      <a
        href="/shelf/authors/dave-draper/"
      >
        Dave Draper
      </a>
    </div>
    <div
      class="progressStatsCss"
    >
      1
       / 
      1
    </div>
  </li>
  <li
    class="listItemCss"
  >
    <div
      class="listItemLinkCss"
    >
      <a
        class="listItemLinkCss"
        href="/shelf/authors/stephen-king/"
      >
        <svg
          class="listItemAvatarCss defaultImageCss"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16 8A8 8 0 110 8a8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zM8 9a5 5 0 00-4.546 2.916A5.986 5.986 0 008 14a5.986 5.986 0 004.546-2.084A5 5 0 008 9z"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          class="progressRingCss"
          preserveAspectRatio="none"
          viewBox="0 0 36 36"
        >
          <g
            id="circles"
            stroke-width="1"
          >
            <circle
              class="percentBackgroundCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
            />
            <circle
              class="percentProgressCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
              stroke-dasharray="109.95574287564276 109.95574287564276"
              stroke-dashoffset="0"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="listItemTitleCss"
    >
      <a
        href="/shelf/authors/stephen-king/"
      >
        Stephen King
      </a>
    </div>
    <div
      class="progressStatsCss"
    >
      1
       / 
      1
    </div>
  </li>
  <li
    class="listItemCss"
  >
    <div
      class="listItemLinkCss"
    >
      <a
        class="listItemLinkCss"
        href="/shelf/authors/richard-laymon/"
      >
        <svg
          class="listItemAvatarCss defaultImageCss"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16 8A8 8 0 110 8a8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zM8 9a5 5 0 00-4.546 2.916A5.986 5.986 0 008 14a5.986 5.986 0 004.546-2.084A5 5 0 008 9z"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          class="progressRingCss"
          preserveAspectRatio="none"
          viewBox="0 0 36 36"
        >
          <g
            id="circles"
            stroke-width="1"
          >
            <circle
              class="percentBackgroundCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
            />
            <circle
              class="percentProgressCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
              stroke-dasharray="109.95574287564276 109.95574287564276"
              stroke-dashoffset="0"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="listItemTitleCss"
    >
      <a
        href="/shelf/authors/richard-laymon/"
      >
        Richard Laymon
      </a>
    </div>
    <div
      class="progressStatsCss"
    >
      1
       / 
      1
    </div>
  </li>
  <li
    class="listItemCss"
  >
    <div
      class="listItemLinkCss"
    >
      <a
        class="listItemLinkCss"
        href="/shelf/authors/brad-schoenfeld/"
      >
        <svg
          class="listItemAvatarCss defaultImageCss"
          fill="currentColor"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M16 8A8 8 0 110 8a8 8 0 0116 0zm-6-3a2 2 0 11-4 0 2 2 0 014 0zM8 9a5 5 0 00-4.546 2.916A5.986 5.986 0 008 14a5.986 5.986 0 004.546-2.084A5 5 0 008 9z"
            fill-rule="evenodd"
          />
        </svg>
        <svg
          class="progressRingCss"
          preserveAspectRatio="none"
          viewBox="0 0 36 36"
        >
          <g
            id="circles"
            stroke-width="1"
          >
            <circle
              class="percentBackgroundCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
            />
            <circle
              class="percentProgressCss"
              cx="18"
              cy="18"
              fill="none"
              r="17.5"
              stroke-dasharray="109.95574287564276 109.95574287564276"
              stroke-dashoffset="0"
            />
          </g>
        </svg>
      </a>
    </div>
    <div
      class="listItemTitleCss"
    >
      <a
        href="/shelf/authors/brad-schoenfeld/"
      >
        Brad Schoenfeld
      </a>
    </div>
    <div
      class="progressStatsCss"
    >
      1
       / 
      1
    </div>
  </li>
</ul>
`;
